tests/Unit/Utilities/Test_Functional.cpp

```diff
@@ -0,0 +1,232 @@
+
+#define MAKE_UNARY_TEST(STRUCTNAME, FUNC)                         \
+  template <typename T, typename UniformGen, typename RandomDist> \
+  void test_func_eval_##FUNC(UniformGen gen, RandomDist dist) {   \
+    T val = dist(gen);                                            \
+    CHECK(STRUCTNAME(val) == FUNC(val));                          \
+  }
+
+#define MAKE_BINARY_TEST(STRUCTNAME, FUNC)                      \
+  template <typename T1, typename T2, typename UniformGen,      \
+            typename RandomDist1, typename RandomDist2>         \
+  void test_func_eval_##FUNC(UniformGen gen, RandomDist1 dist1, \
+                             RandomDist2 dist2) {               \
+    T1 val1 = dist1(gen);                                       \
+    T2 val2 = dist2(gen);                                       \
+    CHECK(STRUCTNAME(val1, val2) == FUNC(val1, val2));          \
+  }
+
+#define MAKE_BINARY_INPLACE_TEST(FUNCNAME, STRUCTNAME, FUNC)        \
+  template <typename T1, typename T2, typename UniformGen,          \
+            typename RandomDist1, typename RandomDist2>             \
+  void test_func_eval_##FUNCNAME(UniformGen gen, RandomDist1 dist1, \
+                                 RandomDist2 dist2) {               \
+    T1 val1 = dist1(gen);                                           \
+    T2 val2 = dist2(gen);                                           \
+    auto result = FUNC(val1, val2);                                 \
+    CHECK(STRUCTNAME(val1, val2) == result);                        \
+    CHECK(val1 == result);                                          \
+  }
+
+MAKE_UNARY_TEST(Abs<>{}, abs)
+MAKE_UNARY_TEST(ACos<>{}, acos)
+MAKE_UNARY_TEST(ACosh<>{}, acosh)
+MAKE_UNARY_TEST(ASin<>{}, asin)
+MAKE_UNARY_TEST(ASinh<>{}, asinh)
+MAKE_UNARY_TEST(ATan<>{}, atan)
+MAKE_UNARY_TEST(ATanh<>{}, atanh)
+MAKE_UNARY_TEST(Cbrt<>{}, cbrt)
+MAKE_UNARY_TEST(Cos<>{}, cos)
+MAKE_UNARY_TEST(Cosh<>{}, cosh)
+MAKE_UNARY_TEST(Erf<>{}, erf)
+MAKE_UNARY_TEST(Exp<>{}, exp)
+MAKE_UNARY_TEST(Exp10<>{}, exp10)
+MAKE_UNARY_TEST(Exp2<>{}, exp2)
+MAKE_UNARY_TEST(FAbs<>{}, fabs)
+MAKE_UNARY_TEST(Imag<>{}, imag)
+MAKE_UNARY_TEST(InvCbrt<>{}, invcbrt)
+MAKE_UNARY_TEST(InvSqrt<>{}, invsqrt)
+MAKE_UNARY_TEST(Log<>{}, log)
+MAKE_UNARY_TEST(Log10<>{}, log10)
+MAKE_UNARY_TEST(Log2<>{}, log2)
+MAKE_UNARY_TEST(Real<>{}, real)
+MAKE_UNARY_TEST(Sin<>{}, sin)
+MAKE_UNARY_TEST(Sinh<>{}, sinh)
+MAKE_UNARY_TEST(StepFunctionStruct<>{}, step_function)
+MAKE_UNARY_TEST(Sqrt<>{}, sqrt)
+MAKE_UNARY_TEST(Tan<>{}, tan)
+MAKE_UNARY_TEST(Tanh<>{}, tanh)
+
+MAKE_BINARY_TEST(ATan2<>{}, atan2)
+MAKE_BINARY_TEST(Hypot<>{}, hypot)
+
+MAKE_BINARY_INPLACE_TEST(divassign, DivAssign<>{}, std::divides<>{});
+MAKE_BINARY_INPLACE_TEST(minusassign, MinusAssign<>{}, std::minus<>{});
+MAKE_BINARY_INPLACE_TEST(multassign, MultAssign<>{}, std::multiplies<>{});
+MAKE_BINARY_INPLACE_TEST(plusassign, PlusAssign<>{}, std::plus<>{});
+
+using RealTypeList = brigand::list<float, double>;
```

Comment:
Why are you using `brigand`? This isn't done anywhere else, you should use `tmpl`